src/pcrlock/pcrlock.c

```diff
@@ -2769,17 +2855,13 @@ static int verb_lock_raw(int argc, char *argv[], void *userdata) {
                         return log_error_errno(errno, "Failed to open '%s': %m", argv[1]);
         }
 
-        r = read_full_stream(f ?: stdin, &data, &size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to read data from stdin: %m");
-
         for (uint32_t i = 0; i < TPM2_PCRS_MAX; i++) {
                 _cleanup_(json_variant_unrefp) JsonVariant *record = NULL;
 
                 if (!FLAGS_SET(arg_pcr_mask, UINT32_C(1) << i))
                         continue;
 
-                r = make_pcrlock_record(i, data, size, &record);
+                r = make_pcrlock_record_from_stream(i, f ?: stdin, &record);
```

Comment:
Can this really be done in a loop? Surely the stream needs to be rewound (not going to work for stdin) if multiple PCRs are specified or this need to enforce that only a single PCR is specified.

Reply:
the code read a buffer update the different digests with the content of the buffer: this avoid re-reading the same content for every hash, and doing a seek(0) (that as you say cannot be applied for stdin)

Reply:
But isn't the digest re-calculated from the stream for each PCR?  Or the PCR loop should be moved to make_pcrlock_record_from_stream (returning the record array rather than a single record)? The initrd case is obviously fine as there is only a single PCR involved.

Reply:
Or alternatively return the digest rather than the record and then take this call outside the loop and build the record in the caller.

Reply:
I am not sure that I follow you.

The loop is about hashes algorithms, not PCRs.  The output should be the hash of the asset (file) for each algorithm.

If there are multiple PCRs, those will share the same digest and the JSON output will contain something like this:

```json
{
  "records": [
    {"pcr": 1, "digests": [ <HASHES> ]},
    {"pcr": 2, "digests": [ <SAME HASHES> ]}
  ]
}
```

That it is stored is the value of the hash that will be used to extend this PCR, not the extended value.

Reply:
@qhill is right: there *is* a loop around PCRs.

Reply:
in verb_lock_raw() we iterate through arg_pcr_mask...

in verb_lock_kernel_initrd() we only do this for a single PCR, hence there we are fine.

Reply:
hence, please change make_pcrlock_record_from_stream() to take a pcr mask, not a single pcr index, and things should be good.

Reply:
oh I think that I get it now, is the 2776 loop!

Reply:
(and please don't try to rewind the stream. that'd suck, since the tool wouldn#t work on pipes anymore, and would mean we do the same work over and over again)

Reply:
@qhill sorry for the delay. I think that I addressed the iteration over the PCRs issue.  As recommended by @poettering I used a `pcr_mask` parameter.

@qhill thanks a lot for pointing this bug! Can you check if this change makes sense?

Reply:
No problem.  This looks correct now.